File: feature-route/src/main/kotlin/dev/roozbahani/trailmetrics/feature/route/RouteViewModel.kt
package dev.roozbahani.trailmetrics.feature.route

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dev.roozbahani.trailmetrics.core.error.RouteUiError
import dev.roozbahani.trailmetrics.core.error.RouteUiErrorMapper
import dev.roozbahani.trailmetrics.domain.model.ActivityType
import dev.roozbahani.trailmetrics.domain.model.Coordinates
import dev.roozbahani.trailmetrics.domain.model.Route
import dev.roozbahani.trailmetrics.domain.model.RouteDraft
import dev.roozbahani.trailmetrics.domain.model.RouteError
import dev.roozbahani.trailmetrics.domain.model.RoutePoint
import dev.roozbahani.trailmetrics.domain.model.UserProfile
import dev.roozbahani.trailmetrics.domain.repository.UserProfileRepository
import dev.roozbahani.trailmetrics.domain.usecase.GenerateClosedRouteUseCase
import dev.roozbahani.trailmetrics.domain.usecase.GetCurrentLocationUseCase
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class RouteViewModel(
    private val getCurrentLocationUseCase: GetCurrentLocationUseCase,
    private val generateClosedRouteUseCase: GenerateClosedRouteUseCase,
    private val userProfileRepository: UserProfileRepository,
    private val uiErrorMapper: RouteUiErrorMapper
) : ViewModel() {

    private val _uiState = MutableStateFlow(RouteUiState())
    val uiState: StateFlow<RouteUiState> = _uiState.asStateFlow()

    private val _uiEvents = Channel<RouteUiEvent>(Channel.BUFFERED)
    val uiEvents: Flow<RouteUiEvent> = _uiEvents.receiveAsFlow()

    init {
        loadCurrentLocation()
        getAndUpdateUserProfile()
    }

    private fun loadCurrentLocation() {
        viewModelScope.launch {
            getCurrentLocationUseCase()
                .onSuccess { coordinates ->
                    _uiState.update { it.copy(startPoint = coordinates) }
                }
                .onFailure { throwable ->
                    handleCurrentLocationErrors(throwable as? RouteError)
                }
        }
    }

    private fun handleCurrentLocationErrors(error: RouteError?) {
        viewModelScope.launch {
            _uiEvents.send(RouteUiEvent.ShowError(error = uiErrorMapper.map(error)))

            if (error is RouteError.MissingLocationPermission) {
                _uiEvents.send(RouteUiEvent.RequestLocationPermission)
            }
        }
    }

    fun onLocationPermissionGranted() {
        loadCurrentLocation()
    }

    fun onMapTapped(coordinates: Coordinates) {
        _uiState.update { state ->
            val newWaypoint = RoutePoint(coordinates = coordinates, order = state.waypoints.size)
            state.copy(waypoints = state.waypoints + newWaypoint)
        }
    }

    fun onGenerateRouteClicked() {
        val state = _uiState.value
        val startPoint = state.startPoint ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            val draftRoute = RouteDraft(
                startPoint = RoutePoint(coordinates = startPoint, order = 0),
                waypoints = state.waypoints
            )

            generateClosedRouteUseCase(draftRoute)
                .onSuccess { route ->
                    _uiState.update { it.copy(isLoading = false, generatedRoute = route) }
                }
                .onFailure { error ->
                    _uiState.update { it.copy(isLoading = false) }
                    _uiEvents.send(RouteUiEvent.ShowError(uiErrorMapper.map(error as? RouteError)))
                }
        }
    }

    fun onResetClicked() {
        _uiState.update { RouteUiState() }
        loadCurrentLocation()
    }

    fun onWaypointRemoved(removeCandidate: RoutePoint) {
        _uiState.update { state ->
            val updatedWaypoints = state.waypoints
                .filterNot { it == removeCandidate }
                .mapIndexed { index, point -> point.copy(order = index) }

            state.copy(waypoints = updatedWaypoints, generatedRoute = null)
        }
    }

    fun onActivityTypeSelected(activityType: ActivityType) {
        _uiState.update { state ->
            state.copy(selectedActivityType = activityType)
        }
    }

    fun saveUserProfile(weightKg: Double) {
        viewModelScope.launch {
            val userProfile = UserProfile(weightKg)
            userProfileRepository.saveUserProfile(userProfile)
            _uiState.update { state -> state.copy(userProfile = userProfile) }
        }
    }

    fun onStartTrackingClicked() {
        viewModelScope.launch {
            val profile = userProfileRepository.getUserProfile()
            if (profile == null) {
                _uiEvents.send(RouteUiEvent.RequestUserProfile)
            } else {
                val startPoint = uiState.value.startPoint
                val plannedRoutePoints = uiState.value.generatedRoute?.points?.map { it.coordinates }
                if (startPoint != null && !plannedRoutePoints.isNullOrEmpty()) {
                    _uiEvents.send(
                        RouteUiEvent.NavigateToTracking(
                            startPoint = startPoint,
                            plannedRoutePoints = plannedRoutePoints,
                            activityType = uiState.value.selectedActivityType
                        )
                    )
                }
            }
        }
    }

    private fun getAndUpdateUserProfile() {
        viewModelScope.launch {
            val userProfile = userProfileRepository.getUserProfile()
            _uiState.update { state -> state.copy(userProfile = userProfile) }
        }
    }
}

data class RouteUiState(
    val startPoint: Coordinates? = null,
    val waypoints: List<RoutePoint> = emptyList(),
    val generatedRoute: Route? = null,
    val userProfile: UserProfile? = null,
    val selectedActivityType: ActivityType = ActivityType.Running, // by default
    val isLoading: Boolean = false
) {
    val canGenerateRoute: Boolean
        get() = startPoint != null && waypoints.size >= MIN_WAYPOINTS && !isLoading

    private companion object {
        const val MIN_WAYPOINTS = 3
    }
}

sealed interface RouteUiEvent {
    data object RequestLocationPermission : RouteUiEvent
    data class ShowError(val error: RouteUiError) : RouteUiEvent
    data object RequestUserProfile : RouteUiEvent
    data class NavigateToTracking(
        val startPoint: Coordinates,
        val plannedRoutePoints: List<Coordinates>,
        val activityType: ActivityType
    ) : RouteUiEvent
}

File: iosApp/Packages/Route/Sources/Route/RouteViewModel.swift
//
//  RouteViewModel.swift
//  Route
//

import Foundation
import SharedKit

@MainActor
public class RouteViewModel: ObservableObject {
    private static let minWaypoints = 3

    @Published public var startPoint: Coordinates?
    @Published public var waypoints: [RoutePoint] = []
    @Published public var generatedRoute: Route?
    @Published public var userProfile: UserProfile?
    @Published public var selectedActivityType: ActivityType = .running
    @Published public var isLoading = false

    public var canGenerateRoute: Bool {
        startPoint != nil && waypoints.count >= Self.minWaypoints && !isLoading
    }

    // Not a stored `let events` created once in init(): AsyncStream supports only one
    // live consumer over its lifetime. RouteView's `.task` consuming this is cancelled
    // whenever a pushed screen (e.g. Tracking) covers it, and gets a brand-new `Task`
    // when that screen pops and RouteView is revealed again — but a second `for await`
    // over an AsyncStream whose first consumer was cancelled (rather than the stream
    // finishing on its own) exits immediately without ever receiving anything, silently.
    // Exposing a factory instead lets each fresh `.task` get its own fresh stream.
    private var eventsContinuation: AsyncStream<RouteUiEvent>.Continuation?

    private let getCurrentLocationUseCase: GetCurrentLocationUseCase
    private let generateClosedRouteUseCase: GenerateClosedRouteUseCase
    private let userProfileRepository: UserProfileRepository

    public init(
        getCurrentLocationUseCase: GetCurrentLocationUseCase = KoinHelper().getCurrentLocationUseCase(),
        generateClosedRouteUseCase: GenerateClosedRouteUseCase = KoinHelper().generateClosedRouteUseCase(),
        userProfileRepository: UserProfileRepository = KoinHelper().userProfileRepository()
    ) {
        self.getCurrentLocationUseCase = getCurrentLocationUseCase
        self.generateClosedRouteUseCase = generateClosedRouteUseCase
        self.userProfileRepository = userProfileRepository

        loadCurrentLocation()
        getAndUpdateUserProfile()
    }

    public func makeEventsStream() -> AsyncStream<RouteUiEvent> {
        let (stream, continuation) = AsyncStream.makeStream(of: RouteUiEvent.self)
        eventsContinuation = continuation
        return stream
    }

    private func emit(_ event: RouteUiEvent) {
        eventsContinuation?.yield(event)
    }

    private func loadCurrentLocation() {
        Task {
            do {
                startPoint = try await getCurrentLocationUseCase.invoke()
            } catch {
                emit(.showError(error.underlyingRouteError?.toUiError() ?? RouteUiErrorGeneral.shared))
            }
        }
    }

    public func onMapTapped(_ coordinates: Coordinates) {
        let waypoint = RoutePoint(coordinates: coordinates, order: Int32(waypoints.count))
        waypoints.append(waypoint)
    }

    public func onGenerateRouteClicked() {
        guard let startPoint else { return }

        isLoading = true

        let draft = RouteDraft(
            startPoint: RoutePoint(coordinates: startPoint, order: 0),
            waypoints: waypoints
        )

        Task {
            do {
                generatedRoute = try await generateClosedRouteUseCase.invoke(draft: draft)
                isLoading = false
            } catch {
                isLoading = false
                emit(.showError(error.underlyingRouteError?.toUiError() ?? RouteUiErrorGeneral.shared))
            }
        }
    }

    public func onResetClicked() {
        startPoint = nil
        waypoints = []
        generatedRoute = nil
        userProfile = nil
        selectedActivityType = .running
        isLoading = false
        loadCurrentLocation()
    }

    public func onWaypointRemoved(_ removeCandidate: RoutePoint) {
        let updatedWaypoints = waypoints
            .filter { $0 != removeCandidate }
            .enumerated()
            .map { index, point in RoutePoint(coordinates: point.coordinates, order: Int32(index)) }

        waypoints = updatedWaypoints
        generatedRoute = nil
    }

    public func onActivityTypeSelected(_ activityType: ActivityType) {
        selectedActivityType = activityType
    }

    public func saveUserProfile(weightKg: Double) {
        Task {
            let profile = UserProfile(weightKg: weightKg)
            try await userProfileRepository.saveUserProfile(userProfile: profile)
            userProfile = profile
        }
    }

    public func onStartTrackingClicked() {
        Task {
            guard try await userProfileRepository.getUserProfile() != nil else {
                emit(.requestUserProfile)
                return
            }

            guard let startPoint, let generatedRoute, !generatedRoute.points.isEmpty else {
                return
            }

            emit(
                .navigateToTracking(
                    startPoint: startPoint,
                    plannedRoutePoints: generatedRoute.points.map { $0.coordinates },
                    activityType: selectedActivityType
                )
            )
        }
    }

    private func getAndUpdateUserProfile() {
        Task {
            userProfile = try await userProfileRepository.getUserProfile()
        }
    }
}

private extension Error {
    /// SKIE bridges a thrown Kotlin exception to an NSError with the real value in userInfo["KotlinException"].
    var underlyingRouteError: RouteError? {
        guard let nsError = self as? NSError,
              let kotlinException = nsError.userInfo["KotlinException"] else {
            return nil
        }
        return kotlinException as? RouteError
    }
}
